### PR TITLE
 persist: add allow_compaction() operator

### PR DESCRIPTION
**Pull request by aljoscha**
This operator will be used to allow compaction for persisted sources.

The rustdoc and comments have all the information.

### Checklist

The first commit is from #9938, and makes testing simpler. Without it I would have to first seal up to some frontier beyond my testing compaction frontiers, which would add chaff to the tests.

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/persist/src/operators/stream.rs [hunk 1]
```diff
@@ -463,6 +466,200 @@ where
...
+        // (that is >, not >=) this candidate, it becomes the compaction frontier.
+        //
+        // NOTE: We could try and be more clever here, say we could apply the logic that checks if
+        // we moved beyond a given frontier only to the input frontier, not the combined/effective
+        // frontier, and move through candidates like this. The reasoning behind this is that we
+        // don't have to wait until the allowed frontier advances past a given frontier, we only
+        // need to be careful not to advance the compaction frontier up to the seal frontier, for
+        // which the input frontier is a proxy.
+        let mut candidate_compaction_frontier: Antichain<u64> =
+            Antichain::from_elem(timely::progress::Timestamp::minimum());
```
Comment (danhhz) [inline src/persist/src/operators/stream.rs:556]: nit: add a use for timely::progress::Timestamp at the top of the file

File: src/persist/src/operators/stream.rs [hunk 2]
```diff
@@ -463,6 +466,200 @@ where
...
+                    candidate_compaction_frontier.extend(effective_input_frontier.iter().cloned());
+
+                    tracing::trace!(
+                        "In {}, allowing compaction up to {:?}. Effective input frontier: {:?}",
+                        &operator_name,
+                        compaction_frontier,
+                        effective_input_frontier
+                    );
+
+                    let future = write.allow_compaction(compaction_frontier.clone());
```
Comment (danhhz) [inline src/persist/src/operators/stream.rs:604]: super nit: future is a common module name. I've seen petros use "fut" for this, which I like

File: src/persist/src/operators/stream.rs [hunk 3]
```diff
@@ -463,6 +466,200 @@ where
...
+
+                    pending_futures.push_back(CompactionFuture {
+                        frontier: compaction_frontier.clone(),
+                        future,
+                    });
+                }
+
+                // Swing through all pending futures and see if they're ready. Ready futures will
```
Comment (danhhz) [inline src/persist/src/operators/stream.rs:612]: Not for this PR, but really seems like a bunch of these operator impls have a pretty common pattern that we could maybe extract.

**Review by danhhz — APPROVED** [review]
Very good, I like this much better than the -1!

**Comment by aljoscha** [comment]
tftr! I incorporated the nits.

**Pull request merged**